### PR TITLE
Faster initial load

### PR DESCRIPTION
**Pull request by starcalibre**
Here are the changes from today.
* Update the UI to use the feature document to get the distribution of a feature. This prevents the need for the UI to download the whole dataset when it's initially loaded.
* Update the readme to describe the feature document, and added some info about which indices to add to the DB.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -52,6 +52,11 @@ MongoDB Schema
 Documents should by default be stored in the the database ``microscopium``,
 however this can be changed by updating the ``config/db.js`` file.
 
+### Document Schemas ###
+
+While MongoDB does not enforce a schema for data, certain fields must be present in
+each document to ensure proper operation of the UI.
+
```
Comment (jni) [inline README.md]: Ooh I like this paragraph! =)

File: app/routes.js [hunk 2]
```diff
@@ -24,21 +28,14 @@ module.exports = function(app) {
...
     });
 
-    // get all screen ids
-    app.get('/api/screen_ids', function(req, res) {
+    // get fields of all screens
+    app.get('/api/screens/:field/', function(req, res) {
         Screen
             .find({})
-            .select('_id')
+            .select(req.params.field)
```
Comment (jni) [inline app/routes.js]: What's a field?

Comment (jni) [inline app/routes.js]: definitely needs clarification — can only think of microscope tiles, which I don't think is what you're talking about here!

File: app/routes.js [hunk 3]
```diff
@@ -46,43 +43,50 @@ module.exports = function(app) {
...
-    app.get('/api/sample/:id', function(req, res) {
-          Sample
-            .find({ '_id': req.params.id })
-            .exec(resHandler(res))
+    // get value from sample document
+    app.get('/api/sample/:id/:field', function(req, res) {
+      Sample
+        .find({ '_id': req.params.id })
+        .select(req.params.field)
+        .exec(resHandler(res))
```
Comment (jni) [inline app/routes.js]: Hey look, these are exactly the same lines as 49-51! -> Refactor. =) I see that "sample" and "screen" are different, but you can parameterise on that. Approximate example:

```javascript
get_field_callback = function(collection) {
    return function(req, res) {
        collection.find({}).select().exec() }

app.get('/api/sample/:id/:field', get_field_callback(Sample))`
```

File: app/routes.js [hunk 4]
```diff
@@ -46,43 +43,50 @@ module.exports = function(app) {
...
-                .exec(function(err, data) {
+                if(req.query.pluck) {
+                    if(err) {
+                        res.send(err);
+                    }
+                    var pluck = _.deepPluck(data,
+                        [req.params.field, req.query.pluck]);
+                    res.send(pluck);
+                }
+                else {
```
Comment (jni) [inline app/routes.js]: Baaaah you're killing me! Anon functions three levels deep!!!! =(

File: public/js/Lineplot.js [hunk 5]
```diff
@@ -214,13 +201,14 @@ Lineplot.prototype.keypressUpdate = function(keyCode) {
     var self = this;
 
     if(self.activeFeature) {
-        if(keyCode === RIGHTARROW && self.activeFeature < this.featureLength) {
+        if(keyCode === RIGHTARROW &&
+            self.activeFeature < this.featureVector.length) {
```
Comment (jni) [inline public/js/Lineplot.js]: Is this 0- or 1-indexed? If 0, won't this fail when we get to the rightmost feature and try to increment it?

**Comment by jni** [comment]
Done with comments! I'm going to be a pain and ask for fewer anon functions where it is at all possible!

**Comment by jni** [comment]
Oh one more comment: at some point (after I've cleaned up feature computation for problematic images), instead of loading the full feature distribution, we can reduce it even further by loading some distribution descriptors, e.g. all the percentiles, or just a histogram with 100 bins. This can get added here or just as an issue.

File: public/js/Lineplot.js (hunk 5 above)
Comment (starcalibre) [inline public/js/Lineplot.js]: The array of feature vectors is 0-indexed, but in the plot they're 1-indexed.. i.e. feature 0 maps to the x=1 co-ordinate of the line plot, feature 1 maps to x=2 etc. 

I had to think about this for more than 30 seconds, so I better add a comment explaining the indexing !

File: app/routes.js (hunk 2 above)
Comment (starcalibre) [inline app/routes.js]: It's a field in the context of database.. i.e. a column name. This is what they're referred to in the MongoDB documentation. I think they avoid the use of the word "key" as that usually refers to unique identifiers such as ``_id``.

File: app/routes.js (hunk 3 above)
Comment (starcalibre) [inline app/routes.js]: I've created issue #52 for this !

Comment (jni) [inline app/routes.js]: @starcalibre are you saying you're punting on not displeasing me with the huge number of anons? =)

Incidentally, #52 becomes much, much easier if you create functions (only need to modify the function) vs using anons (need to modify each and every anon).

**Comment by jni** [comment]
@starcalibre let me know when/if you think this is ready for another round of revision. If you push me I'll merge with the anons in there, but I consider it technical debt and I think we have quite a bit of it already! (Search "US government shutdown 2014" for examples of how this can escalate... =P)

**Comment by starcalibre** [comment]
@jni Hey I'm in the middle of rewriting the API, routes and AJAX request accordingly. This. is. looking. AWESOME. It'll be a substantial bunch of commits though.. would it be easier for you look over if it was in a separate PR?

**Comment by jni** [comment]
Nah, just pile it on to this one — it'll make me happier. =)

**Comment by jni** [comment]
btw, sounds very exciting! =D

**Comment by starcalibre** [comment]
@jni Okay, this is good to go! I think the comments and code make it fairly obvious where I'm going with it but let me know if anything isn't clear :)

The DB API still has lots more functionality to add, such as support for more sophisticated queries but this is a good foundation for now.

**Comment by jni** [comment]
@starcalibre I've made some comments on the individual commits (easier to review). This is *way* better! Thanks! (And great new docs!) Only a couple of minor things and it's going in. Brilliant! =)

**Comment by starcalibre** [comment]
@jni This should be good to go !

**Comment by jni** [comment]
Music to my eyeballs! ... Or something.

Looks wonderful. I'm very happy with the refactor. In it goes!

**Pull request merged**